File: src/main/java/de/com/fdm/tmi/Reader.java
package de.com.fdm.tmi;

import com.github.twitch4j.chat.TwitchChat;
import com.github.twitch4j.chat.TwitchChatBuilder;
import com.github.twitch4j.chat.events.channel.ChannelMessageActionEvent;
import com.github.twitch4j.chat.events.channel.ChannelMessageEvent;
import de.com.fdm.redis.RedisService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Queue;
import java.util.Set;
import java.util.concurrent.ArrayBlockingQueue;

@Component
public class Reader {
    private static final Logger LOG = LoggerFactory.getLogger(Reader.class);
    // if this many messages don't get picked up by a client, the channel gets parted
    private final int inactivityThreshold;
    private final TwitchChat twitchChat;
    private final RedisService redisService;

    private final HashMap<String, Queue<Long>> channelUsageHistory;

    @Autowired
    public Reader(RedisService redisService, @Value("${inactivityThreshold}") int inactivityThreshold) {
        this.redisService = redisService;
        this.inactivityThreshold = inactivityThreshold;
        this.channelUsageHistory = new HashMap<>();

        twitchChat = TwitchChatBuilder.builder().build();
        twitchChat.getEventManager().onEvent(ChannelMessageEvent.class, this::handleMessage);
        twitchChat.getEventManager().onEvent(ChannelMessageActionEvent.class, this::handleActionMessage);
    }

    private void handleMessage(ChannelMessageEvent event) {
        TwitchMessageDto twitchMessageDto = TwitchMessageDto.fromEvent(event);
        String channel = twitchMessageDto.channel();

        if (!isChannelUsed(channel)) {
            leaveChannel(channel);
            LOG.info("Left channel #{} due to inactivity.", channel);
            return;
        }

        long numClients = redisService.publishMessage(twitchMessageDto);

        if (channelUsageHistory.get(channel).size() == inactivityThreshold) {
            channelUsageHistory.get(channel).remove();
        }

        channelUsageHistory.get(channel).add(numClients);
    }

    private boolean isChannelUsed(String channel) {
        if (channelUsageHistory.get(channel).size() != inactivityThreshold) {
            return true;
        }

        for (Long entry : channelUsageHistory.get(channel)) {
            if (entry != 0) {
                return true;
            }
        }
        return false;
    }

    private void handleActionMessage(ChannelMessageActionEvent event) {
        ChannelMessageEvent messageEvent = new ChannelMessageEvent(
                event.getChannel(),
                event.getMessageEvent(),
                event.getUser(),
                event.getMessage(),
                event.getPermissions()
        );
        this.handleMessage(messageEvent);
    }

    // twitch4j handles joining the same chat multiple times
    // the join gets ignored in that case!
    public void joinChannel(String channel) {
        channelUsageHistory.put(channel, new ArrayBlockingQueue<>(inactivityThreshold));
        redisService.addChannel(channel);
        twitchChat.joinChannel(channel);

        LOG.info("Joined channel #{}.", channel);
    }

    private void leaveChannel(String channel) {
        redisService.removeChannel(channel);
        twitchChat.leaveChannel(channel);
    }

    public void joinSavedChannels() {
        Set<String> channels = redisService.getChannels();

        for (String channel : channels) {
            joinChannel(channel);
        }
    }
}

File: src/main/java/de/com/fdm/tmi/Reader.kt
package de.com.fdm.tmi

import com.github.twitch4j.chat.TwitchChat
import com.github.twitch4j.chat.TwitchChatBuilder
import com.github.twitch4j.chat.events.channel.ChannelMessageActionEvent
import com.github.twitch4j.chat.events.channel.ChannelMessageEvent
import de.com.fdm.redis.RedisListener
import de.com.fdm.redis.RedisService
import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Timer
import org.redisson.Redisson
import org.redisson.api.RedissonClient
import org.redisson.config.Config
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.io.File
import java.util.Queue
import java.util.concurrent.ArrayBlockingQueue

@Component
class Reader @Autowired constructor(
    private val redisService: RedisService,
    meterRegistry: MeterRegistry,
    @param:Value("\${inactivityThreshold}") private val inactivityThreshold: Int
) {
    private val log = LoggerFactory.getLogger(RedisListener::class.java)
    private val client: RedissonClient
    private val twitchChat: TwitchChat
    private val channelUsageHistory = HashMap<String, Queue<Long>>()
    private val timer = Timer.builder("tmiReceiver.handleMessage.timer")
                                    .description("Times how long it takes to handle message received from twitch")
                                    .register(meterRegistry)

    init {
        val config = Config.fromYAML(File("src/main/resources/redisson_config.yaml"))
        this.client = Redisson.create(config)

        this.twitchChat = TwitchChatBuilder.builder().build()
        twitchChat.eventManager.onEvent(ChannelMessageEvent::class.java) {
                event: ChannelMessageEvent -> timer.record { handleMessage(event) }
        }

        twitchChat.eventManager.onEvent(ChannelMessageActionEvent::class.java) {
                event: ChannelMessageActionEvent -> handleActionMessage(event)
        }
    }

    private fun handleMessage(event: ChannelMessageEvent) {
        val twitchMessageDto = TwitchMessageDto(event)
        val channel = twitchMessageDto.channel
        if (!isChannelUsed(channel)) {
            log.info("Channel #{} is not used.", channel)
            leaveChannel(channel)
            return
        }
        val numClients = redisService.publishMessage(twitchMessageDto)
        if (channelUsageHistory[channel]!!.size == inactivityThreshold) {
            channelUsageHistory[channel]!!.remove()
        }
        channelUsageHistory[channel]!!.add(numClients)
    }

    private fun handleActionMessage(event: ChannelMessageActionEvent) {
        val messageEvent = ChannelMessageEvent(
            event.channel,
            event.messageEvent,
            event.user,
            event.message,
            event.permissions
        )
        handleMessage(messageEvent)
    }

    private fun isChannelUsed(channel: String): Boolean {
        if (channelUsageHistory[channel]!!.size != inactivityThreshold) {
            return true
        }
        for (entry in channelUsageHistory[channel]!!) {
            if (entry != 0L) {
                return true
            }
        }
        return false
    }

    // twitch4j handles joining the same chat multiple times
    // the join gets ignored in that case!
    fun joinChannel(channel: String) {
        channelUsageHistory[channel] = ArrayBlockingQueue(inactivityThreshold)
        redisService.addChannel(channel)
        twitchChat.joinChannel(channel)

        log.info("Joined channel #{}.", channel)
    }

    private fun leaveChannel(channel: String) {
        redisService.removeChannel(channel)
        twitchChat.leaveChannel(channel)
        log.info("Left channel #{}.", channel)
    }

    fun joinSavedChannels() {
        val channels = redisService.getChannels()
        for (channel in channels) {
            joinChannel(channel)
        }
    }
}